spec/unit/ssl/state_machine_spec.rb

```diff
@@ -56,6 +60,35 @@
     end
   end
 
+  context 'when locking' do
+    it 'locks the file prior to running the state machine' do
+      expect(Puppet::FileSystem).to receive(:exclusive_open).and_call_original.ordered
+      expect(cert_provider).to receive(:load_cacerts).and_return(cacerts).ordered
+      expect(cert_provider).to receive(:load_crls).and_return(crls).ordered
+
+      machine.ensure_ca_certificates
+    end
+
+    it 'raises an exception if it fails to lock' do
+      allow(Puppet::FileSystem.instance_variable_get(:@impl)).to receive(:sleep)
+
+      Puppet::FileSystem.exclusive_open(Puppet[:ssl_lockfile], 0644, 'w', 0) do |fd|
+        expect {
+          machine.ensure_ca_certificates
+        }.to raise_error(Puppet::Error, /Another puppet instance is already running; exiting/)
+      end
+    end
```

Comment:
Should there be another test in here for timeouts or another expect inside this test for the timeout message?

Reply:
Similar comment as above. The state machine never sees the timeout error. But there is a test to make sure the `FileSystem.exclusively_open` method raises a timeout error with the expected message (which I had to fix in https://github.com/puppetlabs/puppet/pull/7530/commits/b56bcdb5c30648968961ffe941b4084bc440e1a7

Reply:
I am trying to think how you would test the timeout here - seems like you would have to enable a timeout in the test system that is shorter to show that the error hasn't been raised yet?  I don't know how to do this in rspec.

Reply:
@kris-bosland I didn't follow your comment. If the file is locked and we reach the timeout, then `Puppet::FileSystem.exclusively_open` will raise `Timeout::Error`. The state machine will rescue that error and raise `Puppet::Error` indicating that another instance is already running.